### DISCUSSION TITLE
Not enough memory

### DISCUSSION THREAD

**Discussion by Ruslan1541** [astra-toolbox/astra-toolbox#436] [Q&A]
Hello everyone!

I am a new astra user and I have a problem working with big data (the projection size is 1490 x 1490 x 7200). When I try to load these projections as a numpy array and use astra.create_sino to create input data for reconstruction, I get an error about not having enough memory for this (I have 64 GB of RAM). Even with projections of size 1490x1490x1000, I can't do anything. Maybe I'm doing something wrong or someone knows another easier way to initialize the input data?

This is my code:

import astra
import numpy as np
import tifffile as tiff
from glob import glob
import os
import matplotlib.pyplot as plt 

data_names = sorted(glob(f'{data_path}*t4proj_[0-9]*'))
data = np.zeros([len(data_names), img.shape[0], img.shape[1]], dtype=np.float16)

for i in range(len(data_names)):
    data[i,:,:] = tiff.imread(data_names[i])

vol_geom = astra.create_vol_geom(data.shape[0], data.shape[2], data.shape[1])
proj_geom = astra.create_proj_geom('cone', 1.838994, 1.838994, 1408, 1740, np.linspace(0, 2*np.pi, 7201, False), 17.2, 795)

proj_id = astra.create_projector('cuda3d', proj_geom, vol_geom)
sinogram_id, sinogram = astra.create_sino(data, proj_id)  // after this i have a error: Unable to allocate 63.5 GiB for an array with shape (1490, 1490, 7021)

Best regars, Ruslan

**Comment by askorikov** [comment]
Hi! Two comments here:
1. I guess you just want to create an Astra object to input to the reconstruction algorithm. In this case you should use `astra.data3d.create('-proj3d', proj_geom, data)` syntax (see [this](https://github.com/astra-toolbox/astra-toolbox/blob/master/samples/python/s006_3d_data.py)). `astra.create_sino` *computes* a sinogram (projection) given the input volume data.
2. 3D reconstructions typically require a lot of computational power and are therefore only implemented on GPU. This means that relevant memory in your case is GPU memory (probably much smaller than 64 GB). For FDK reconstruction, Astra now can automatically split the data into chunks that fit into the available GPU memory, but be aware that the total reconstruction time for your data sizes will probably be significant (tens of minutes or more).

**Reply by Ruslan1541** [reply]
Hi, 
I tried to take your advice and fix my code, but after I run the algorithm (SIRT3D_CUDA), my jupyter notebook kernel just died without using GPU memory. This problem has already appeared before.
In addition, for this case, i used  another  script for reconstructing of parallel-beam projections.

**Reply by askorikov** [reply]
Do you see any error messages? Can you post the new script?

**Reply by askorikov** [reply]
Nevermind the previous questions, I could reproduce the issue. I think you run out of RAM after all, and `astra.data3d.create` doesn't properly catch the failed memory allocation. I will look into that, however this will not help with doing the reconstruction in your case, as you will still not have enough memory. If you have parallel-beam geometry, you can do the reconstruction slice-by-slice, which will dramatically help with memory management. For cone-beam, you will have to implement a more complex chunking of data to fit everything into memory.

**Reply by Ruslan1541** [reply]
Actually, I use data from both beams, but now I'm working more with cone beam data, so this reconstruction is more relevant for me. Can you please provide more information about the implementation of data fragmentation in the case of a cone beam?

**Reply by askorikov** [reply]
I think it would be quite a project to implement larger-than-RAM cone beam reconstruction and I would avoid it if possible. But basically you need to calculate the sizes of sinogram and reconstruction  partitions that can fit into the available RAM, iteratively load these partitions, perform reconstructions, and write them back to the disk. Limiting it to FDK algorithm will almost for sure be necessary, because iterative algorithms will complicate things significantly in terms of ensuring independent partitions.

**Reply by askorikov** [reply]
One optimization that I could suggest is using the 'linking' functionality of Astra (https://astra-toolbox.com/apiref/data3d.html#astra.data3d.link) to avoid creating an internal copy of the Numpy array. This would basically give you a 100% improvement in RAM utilization.
